### ISSUE TITLE
TypeScript Rewrite

### ISSUE THREAD

**Issue by hipstersmoothie** [jimp-dev/jimp#1189]
The types for jimp work but they are separate from the code and the code is super dynamic! this dynamic-ness made writing the types a nightmare (see @crutchcorn's efforts).

We should rewrite jimp to be fully TS and on top of that the typing should be excellent and automatic.

Jimp the package is currently very simple. It requires the plugins that add image type support, the base manipulation function and combines them into a custom jimp. This lets us ship a pre-configured jimp with all the features people would expect but lets our users choose if they in fact want to actually load all those features.

I think our plugin infra can use the typescript "builder" pattern to make the types for a custom jimp instance be really good and really easy.

```ts
const jimp = createJimp()
  .addType(png)
  .addType(jpeg)
  .addType(webp)
  .addPlugin(convolution).
  .addPlugin(canvas)
  .addPlugin(blit)

// Maybe template string to check for support type?
const image = await jimp.read('/my-image.png')

// Methods auto complete
image.canvas().convolution();

await jimp.write('/my-output-image.png')
```

**Comment by crutchcorn** [comment]
It's worth noting that we should probably keep our Type Testing infrastructure in place, even with TS-first code. While most of it can be caught by a build step, passing generics and such still need edgecase handling and more.

**Comment by hipstersmoothie** [comment]
I agree. Their readme does say that we should probably be using https://github.com/SamVerschueren/tsd though

**Comment by marvinruder** [comment]
By the way, I was unable to get types working with [Yarn PnP](https://yarnpkg.com/features/pnp) and the [Yarn Typescript/VSCode SDK](https://yarnpkg.com/getting-started/editor-sdks#tools-currently-supported). I am not sure if this is an issue with my Yarn setup or the rather unusual way types are set up in Jimp. Might be worth looking into during this kind of rewrite. 

<img width="984" alt="image" src="https://user-images.githubusercontent.com/18495294/217205330-6432669b-a305-411a-9c31-186e4aa2560c.png">

**Comment by crutchcorn** [comment]
Truth be told, I don't think that's a bug with our system @marvinruder. It absolutely could be, but I'd be a bit surprised. 

Regardless, if it is an issue, we should be fixing it outside of the TS rewrite. Mind opening a new issue for it and creating a repro repo we could use to debug?

**Comment by marvinruder** [comment]
Will try!

**Comment by hipstersmoothie** [comment]
:rocket: Issue was released in [`v1.0.2`](https://github.com/jimp-dev/jimp/releases/tag/v1.0.2) :rocket:
